File: coordinator/src/shard_manager.rs
use std::collections::HashMap;
use tracing::{info, debug};

#[derive(Debug, Clone)]
pub struct ShardInfo {
    pub shard_id: String,
    pub worker_id: String,
    pub table_name: String,
    pub key_range: Option<(String, String)>,
    pub row_count: u64,
}

#[derive(Debug, Clone)]
pub enum ShardType {
    Hash,
    Range,
    RoundRobin,
}

#[derive(Debug)]
pub struct ShardManager {
    shards: HashMap<String, Vec<ShardInfo>>,
    shard_type: ShardType,
}

impl ShardManager {
    pub fn new() -> Self {
        let mut shards = HashMap::new();
        
        // Initialize with some demo shards
        shards.insert("users".to_string(), vec![
            ShardInfo {
                shard_id: "users_shard_1".to_string(),
                worker_id: "worker1".to_string(),
                table_name: "users".to_string(),
                key_range: Some(("A".to_string(), "M".to_string())),
                row_count: 1000,
            },
            ShardInfo {
                shard_id: "users_shard_2".to_string(),
                worker_id: "worker2".to_string(),
                table_name: "users".to_string(),
                key_range: Some(("N".to_string(), "Z".to_string())),
                row_count: 1200,
            },
        ]);

        shards.insert("orders".to_string(), vec![
            ShardInfo {
                shard_id: "orders_shard_1".to_string(),
                worker_id: "worker1".to_string(),
                table_name: "orders".to_string(),
                key_range: Some(("1".to_string(), "5000".to_string())),
                row_count: 5000,
            },
            ShardInfo {
                shard_id: "orders_shard_2".to_string(),
                worker_id: "worker2".to_string(),
                table_name: "orders".to_string(),
                key_range: Some(("5001".to_string(), "10000".to_string())),
                row_count: 5000,
            },
            ShardInfo {
                shard_id: "orders_shard_3".to_string(),
                worker_id: "worker3".to_string(),
                table_name: "orders".to_string(),
                key_range: Some(("10001".to_string(), "15000".to_string())),
                row_count: 5000,
            },
        ]);

        Self {
            shards,
            shard_type: ShardType::Hash,
        }
    }

    pub fn get_shards_for_table(&self, table_name: &str) -> Vec<ShardInfo> {
        self.shards.get(table_name)
            .cloned()
            .unwrap_or_else(|| {
                debug!("No shards found for table: {}", table_name);
                vec![]
            })
    }

    pub fn get_worker_shards(&self, worker_id: &str) -> Vec<ShardInfo> {
        let mut worker_shards = Vec::new();
        
        for shards in self.shards.values() {
            for shard in shards {
                if shard.worker_id == worker_id {
                    worker_shards.push(shard.clone());
                }
            }
        }
        
        worker_shards
    }

    pub fn add_shard(&mut self, table_name: String, shard: ShardInfo) {
        self.shards.entry(table_name)
            .or_insert_with(Vec::new)
            .push(shard);
    }

    pub fn get_shard_distribution(&self) -> HashMap<String, Vec<String>> {
        let mut distribution = HashMap::new();
        
        for (table_name, shards) in &self.shards {
            let mut workers = Vec::new();
            for shard in shards {
                if !workers.contains(&shard.worker_id) {
                    workers.push(shard.worker_id.clone());
                }
            }
            distribution.insert(table_name.clone(), workers);
        }
        
        distribution
    }

    pub fn get_total_rows(&self) -> u64 {
        self.shards.values()
            .flatten()
            .map(|shard| shard.row_count)
            .sum()
    }
}

File: coordinator/src/main/java/com/distributed/sql/coordinator/ShardManager.java
package com.distributed.sql.coordinator;

import com.distributed.sql.common.models.*;
import com.distributed.sql.common.utils.AppLogger;
import com.distributed.sql.common.utils.Tracer;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Manages data distribution and shard assignments across workers
 */
public class ShardManager {

    private final Map<String, List<ShardInfo>> shards;
    private final Map<String, WorkerInfo> workers;
    private final ShardType shardType;

    public ShardManager() {
        this.shards = new ConcurrentHashMap<>();
        this.workers = new ConcurrentHashMap<>();
        this.shardType = ShardType.HASH;

        initializeDefaultShards();
    }

    private void initializeDefaultShards() {
        // Initialize users table shards
        List<ShardInfo> userShards = Arrays.asList(
                new ShardInfo("users_shard_1", "worker1", "users", "A", "M", 1000),
                new ShardInfo("users_shard_2", "worker2", "users", "N", "Z", 1200));
        shards.put("users", userShards);

        // Initialize orders table shards
        List<ShardInfo> orderShards = Arrays.asList(
                new ShardInfo("orders_shard_1", "worker1", "orders", "1", "5000", 5000),
                new ShardInfo("orders_shard_2", "worker2", "orders", "5001", "10000", 5000),
                new ShardInfo("orders_shard_3", "worker3", "orders", "10001", "15000", 5000));
        shards.put("orders", orderShards);

        // Initialize products table shards
        List<ShardInfo> productShards = Arrays.asList(
                new ShardInfo("products_shard_1", "worker1", "products", "Electronics", "Electronics", 5),
                new ShardInfo("products_shard_2", "worker2", "products", "Appliances", "Appliances", 3),
                new ShardInfo("products_shard_3", "worker3", "products", "Sports", "Accessories", 7));
        shards.put("products", productShards);

        AppLogger.info("Initialized default shard distribution");
    }

    public List<ShardInfo> getShardsForTable(String tableName) {
        return shards.getOrDefault(tableName, new ArrayList<>());
    }

    public List<ShardInfo> getWorkerShards(String workerId) {
        List<ShardInfo> workerShards = new ArrayList<>();

        for (List<ShardInfo> tableShards : shards.values()) {
            for (ShardInfo shard : tableShards) {
                if (shard.getWorkerId().equals(workerId)) {
                    workerShards.add(shard);
                }
            }
        }

        return workerShards;
    }

    public void addShard(String tableName, ShardInfo shard) {
        shards.computeIfAbsent(tableName, k -> new ArrayList<>()).add(shard);
        AppLogger.info("Added shard {} for table {} to worker {}",
                shard.getShardId(), tableName, shard.getWorkerId());
    }

    public Map<String, List<String>> getShardDistribution() {
        Map<String, List<String>> distribution = new HashMap<>();

        for (Map.Entry<String, List<ShardInfo>> entry : shards.entrySet()) {
            String tableName = entry.getKey();
            List<String> workers = new ArrayList<>();

            for (ShardInfo shard : entry.getValue()) {
                if (!workers.contains(shard.getWorkerId())) {
                    workers.add(shard.getWorkerId());
                }
            }

            distribution.put(tableName, workers);
        }

        return distribution;
    }

    public long getTotalRows() {
        return shards.values().stream()
                .flatMap(List::stream)
                .mapToLong(ShardInfo::getRowCount)
                .sum();
    }

    public void registerWorker(String workerId, String address, int port) {
        WorkerInfo workerInfo = new WorkerInfo(workerId, address, port);
        workers.put(workerId, workerInfo);
        AppLogger.info("Registered worker: {} at {}:{}", workerId, address, port);
    }

    public void updateWorkerHeartbeat(String workerId, double cpuUsage, double memoryUsage, int activeQueries) {
        WorkerInfo workerInfo = workers.get(workerId);
        if (workerInfo != null) {
            workerInfo.updateHeartbeat(cpuUsage, memoryUsage, activeQueries);
            AppLogger.debug("Updated heartbeat for worker: {} - CPU: {}%, Memory: {}%, Active queries: {}",
                    workerId, cpuUsage, memoryUsage, activeQueries);
        }
    }

    public List<WorkerInfo> getActiveWorkers() {
        return new ArrayList<>(workers.values());
    }

    public WorkerInfo getWorker(String workerId) {
        return workers.get(workerId);
    }

    public boolean isWorkerHealthy(String workerId) {
        WorkerInfo workerInfo = workers.get(workerId);
        if (workerInfo == null) {
            return false;
        }

        // Consider worker healthy if last heartbeat was within 30 seconds
        long timeSinceLastHeartbeat = System.currentTimeMillis() - workerInfo.getLastHeartbeat();
        return timeSinceLastHeartbeat < 30000;
    }

    public void removeWorker(String workerId) {
        workers.remove(workerId);
        AppLogger.warn("Removed worker: {}", workerId);
    }

    /**
     * Worker information for tracking worker status
     */
    public static class WorkerInfo {
        private final String workerId;
        private final String address;
        private final int port;
        private double cpuUsage;
        private double memoryUsage;
        private int activeQueries;
        private long lastHeartbeat;

        public WorkerInfo(String workerId, String address, int port) {
            this.workerId = workerId;
            this.address = address;
            this.port = port;
            this.lastHeartbeat = System.currentTimeMillis();
        }

        public void updateHeartbeat(double cpuUsage, double memoryUsage, int activeQueries) {
            this.cpuUsage = cpuUsage;
            this.memoryUsage = memoryUsage;
            this.activeQueries = activeQueries;
            this.lastHeartbeat = System.currentTimeMillis();
        }

        // Getters
        public String getWorkerId() {
            return workerId;
        }

        public String getAddress() {
            return address;
        }

        public int getPort() {
            return port;
        }

        public double getCpuUsage() {
            return cpuUsage;
        }

        public double getMemoryUsage() {
            return memoryUsage;
        }

        public int getActiveQueries() {
            return activeQueries;
        }

        public long getLastHeartbeat() {
            return lastHeartbeat;
        }
    }
}
